File: 프로그래머스/3/72413. 합승 택시 요금/합승 택시 요금.js
function solution(n, s, a, b, fares) {
  const INF = Number.MAX_SAFE_INTEGER;

  // dist[i][j]
  const dist = Array.from({ length: n + 1 }, () => Array(n + 1).fill(INF));
  for (let i = 1; i <= n; i++) dist[i][i] = 0;

  // 양방향 간선
  for (const [c, d, f] of fares) {
    if (f < dist[c][d]) {
      dist[c][d] = f;
      dist[d][c] = f;
    }
  }

  // Floyd-Warshall
  for (let k = 1; k <= n; k++) {
    for (let i = 1; i <= n; i++) {
      const ik = dist[i][k];
      if (ik === INF) continue;
      for (let j = 1; j <= n; j++) {
        const nd = ik + dist[k][j];
        if (nd < dist[i][j]) dist[i][j] = nd;
      }
    }
  }

  let ans = INF;
  for (let k = 1; k <= n; k++) {
    const cost = dist[s][k] + dist[k][a] + dist[k][b];
    if (cost < ans) ans = cost;
  }

  return ans;
}

File: 프로그래머스/3/72413. 합승 택시 요금/합승 택시 요금.py
def solution(n, s, a, b, fares):
    INF = 10**18
    dist = [[INF] * (n + 1) for _ in range(n + 1)]
    for i in range(1, n + 1):
        dist[i][i] = 0

    for c, d, f in fares:
        # 양방향
        if f < dist[c][d]:
            dist[c][d] = f
            dist[d][c] = f

    # Floyd-Warshall
    for k in range(1, n + 1):
        dk = dist[k]
        for i in range(1, n + 1):
            if dist[i][k] == INF:
                continue
            ik = dist[i][k]
            di = dist[i]
            for j in range(1, n + 1):
                nd = ik + dk[j]
                if nd < di[j]:
                    di[j] = nd

    ans = INF
    for k in range(1, n + 1):
        cost = dist[s][k] + dist[k][a] + dist[k][b]
        if cost < ans:
            ans = cost

    return ans